llvm/include/llvm/MC/MCMachObjectWriter.h

```diff
@@ -108,8 +108,18 @@ class MachObjectWriter : public MCObjectWriter {
         : Sym(Sym), F(F), MRE(MRE) {}
   };
 
+  struct AddendsSizeAndOffset {
```

Comment:
Hey Steven,

As you can see, yes, this is way more complicated than we first thought, and I cannot just extract the addend value, if you see the code in `AArch64AsmBackend.cpp` you will notice that the addend that is generated by `handleFixup` isn't the final value, it is modified further and there is some more work done if the addend is an instruction which is target specific, so I need to store a bunch of additional information to recreate the data at the offset exaclty.

Reply:
On thinking about this a little longer, there is only one way of doing this a little more efficiently, the current problem lies in ARM and AArch64. The issue is that not only is the Addend value not final till the fixup is being handled, it is also OR'ed or AND'ed to the existing value at the offset. The only way to reduce the amount of stuff being stored is to first apply the fixup, then copy the value and zero it out, it will still require we store the Addend, it's offset, and the size of the addend which will be 13 bytes per Addend, does that sound better?

Reply:
I think we need to store the value after applying the fix up, otherwise I don't see this being a gain.

For the storage needed, you might get away without size field by simply always assume the largest size possible. We are encoding as VBR, which means if the value is small, you don't use extra bits to store it.

Reply:
How does it look now?

Reply:
Does the order matter? It isn't a packed struct so it is going to 16 bytes either way.

Reply:
I think I would still reorder it, so that if we add more fields at the end later, there's a better chance at packing.
Alternatively we could store three separate arrays and maximize packing that way.